src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,402 @@
+    c = idx[first(sel)]
+    fun = last(sel)
+    if length(c) > 3
+        newcol = Symbol(join(_names(idx)[c[1:2]], "_"), "_etc_", funname(fun))
+    else
+        newcol = Symbol(join(_names(idx)[c], "_"), "_", funname(fun))
+    end
+    return c => fun => newcol
+end
+
+function select_transform!(nc::Union{Pair{Int, Pair{ColRename, Symbol}},
+                                     Pair{<:Union{Int, AbstractVector{Int}},
+                                          <:Pair{<:Base.Callable, Symbol}}},
+                           df::DataFrame, newdf::DataFrame,
+                           transformed_cols::Dict{Symbol, Any}, copycols::Bool)
+    transform_spec = last(nc)
+    newname = last(transform_spec)
+    if !isnothing(transformed_cols[newname])
+        @assert !hasproperty(newdf, newname)
+    end
+    col_idx = first(nc)
+    if nc isa Pair{Int, Pair{ColRename, Symbol}}
+        newdf[!, newname] = copycols ? df[:, col_idx] : df[!, col_idx]
+    elseif nc isa Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+        newdf[!, newname] = first(transform_spec).(df[!, col_idx])
+    elseif nc isa Pair{<:AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+        if length(col_idx) == 0
+            newdf[!, newname] = [first(transform_spec)() for _ in axes(df, 1)]
+        else
+            rowiterator = Tables.rows(Tables.columntable(df[!, col_idx]))
+            newdf[!, newname] = map(first(transform_spec),
+                                    Tables.namedtupleiterator(eltype(rowiterator), rowiterator))
+        end
+    else
+        throw(ErrorException("code should never reach this branch"))
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing
+provided that the columns requested in each of them are unique and present in `df`.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Column renaming and transformations are supported.
+The syntax for column renaming is `old_column=>new_column_name`.
+The syntax for column transformations is `old_column=>fun=>new_column_name`.
+`new_column_name` must be a `Symbol`.
+If `old_column` is a `Symbol` or an integer then and `fun` must be callable
+that is applied to each element (row) of `old_column`.
+Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
+will be passed a `NamedTuple` representing each row, holding only
+the columns specified by `old_column`.
+
+Column transfomation syntax also supports a short `old_column=>fun` form, in which
+case `new_column_name` is automatically generated by joining `old_column` with `fun` name
+with `_`.
+
+If more than one argument is passed then requesting duplicate column names in target
+data frame are accepted (e.g. `select(df, :a, :, :a)` is allowed) and they get deduplicated.
```

Comment:
It would. However, actually also `Not` makes sense apart from `:` and regex and some general custom vector (which can be a result of some selection earlier).

The problem is that we inherit this behavior from `All` (the example below is using current master not this PR):
```
julia> using DataFrames

julia> df = DataFrame(a=1,b=2)
1×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │

julia> select(df, :a, 1, :a)
1×1 DataFrame
│ Row │ a     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> df[:, All(:a, 1, :a)]
1×1 DataFrame
│ Row │ a     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │
```

The rule that we could introduce (given your comment) is that we disallow passing two or more times a single index (like `:a` or `1` above) with the same value (but this would make the description longer not shorter).

As a reference - we disallow duplicates in vectors so this is an error:
```
julia> df[:, [:a, :a, :a]]
ERROR: ArgumentError: Elements of Symbol[:a, :a, :a] must be unique
```

In summary - this is a bit not intuitive but I would leave it as is as this is what `All` does.

Reply:
Do we really have to do the same thing as `All` here?

Anyway, are you sure being stricter will make the description longer? If we disallow repeated names as single indices, the difference between selection and rename/transform disappears, which allows removing the next sentence.

Reply:
> If we disallow repeated names as single indices

OK - let us go for it, but then I would prefer to have the same with `All` (I do not think anyone uses `select(df, :a, :a)` in practice). Are we OK to change it also there.

@piever - would you consider making `All` stricter this way also in JuliaDB.jl?

Reply:
Though `All` is documented to take the union of selections, so we would have to define it differently...

Reply:
I know - this would be breaking, but it is not clear to me why `select(df, All(:a, :a))` should be allowed while `select(df, :a, :a)` be disallowed.

Reply:
I think it'd be awkward to make `All` stricter. The reason is that (at least in JuliaDB) one can "mix and match" different selectors. So I can imagine one doing `All(Between(:foo, :bar), r"^baz")`  and it seems excessive to error if there is a column that starts with `baz` between `foo` and `bar`.

Reply:
I don't think there is a strong reason why `select(t, args...)` should be the same as `select(t, All(args...))`. In JuliaDB `All` "flattens" the structure, so that `select(t, (tuple1, tuple2))` would return a table with two columns of `StructArrays`, whereas `select(t, All(tuple1, tuple2))` would return a table with 4 columns.

Reply:
OK. If you feel we should differentiate the behaviors then we can do it. But first I just want to clarify before we move forward:
* `All(Between(:foo, :bar), r"^baz")` would work exactly like you expect. We want to differentiate the behavior between `:a` and `[:a]`, so essentially - make a difference if a selection is an index or a set of indices. The proposal is to disallow duplicates if a single index is passed twice, like `All(:a, :a)`, but flatten as an union sets of values in other cases, so `All([:a], [:a])` or `All([:a], :a)` would be OK.
* In your second example now I see that the crucial thing is that in JuliaDB.jl you do not support selection in the form `select(t, selector1, selector2)`, but currently this is exactly defined to mean `select(t, All(selector1, selector2))`.

Reply:
In JuliaDB there are two different ways to combine selectors, namely `(sel1, sel2, sel3)` (that preserves the "nesting" giving a table where some columns can themselves be tables, or more precisely `StructArrays`) and `All(sel1, sel2, sel3)`. Currently the first option `(sel1, sel2, sel3)` errors if a column is repeated, whereas `All(sel1, sel2, sel3)` does not. Intuitively, `select(df, sel1, sel2)` should actually correspond to `select(df, (sel1, sel2))` rather than `select(df, All(sel1, sel2))`.

I suspect the difference is that DataFrames does not really use "nested tables", so the distinction between `All(sel1, sel2)` and `(sel1, sel2)` is probably less useful here. Maybe an alternative is splatting? e.g. `select(df, args...)` would translate to `select(df, collect(args))` and one could do `select(df, sel1..., sel2...)` if `sel1` and `sel2` are sets of indices. This would reserve the option of considering `select(df, sel1, sel2)` as a nested table in the future?

Regarding changing the meaning of `All`, I suspect that the simplification in docstrings here would induce an equivalent complication of the docstring in `All`

Reply:
This is exactly this - we do not do "nesting" now.
Given the discussion I will drop referring in `select` to `All` and just make `select` use its own rules (and `All` will mean what it means now).

Thank you!